xdsl/context.py

```diff
@@ -0,0 +1,228 @@
+from collections.abc import Callable, Iterable
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING
+
+from xdsl.dialects import get_all_dialects
+
+if TYPE_CHECKING:
+    from xdsl.ir import Attribute, Dialect, Operation
+
+
+@dataclass
+class MLContext:
+    """Contains structures for operations/attributes registration."""
+
+    allow_unregistered: bool = field(default=False)
+
+    _loaded_dialects: "dict[str, Dialect]" = field(default_factory=dict)
+    _loaded_ops: "dict[str, type[Operation]]" = field(default_factory=dict)
+    _loaded_attrs: "dict[str, type[Attribute]]" = field(default_factory=dict)
+    _registered_dialects: "dict[str, Callable[[], Dialect]]" = field(
+        default_factory=dict
+    )
+    """
+    A dictionary of all registered dialects that are not yet loaded. This is used to
+    only load the respective Python files when the dialect is actually used.
+    """
+
+    def register_all_dialects(self):
+        """
+        Register all dialects that can be used.
+
+        Add other/additional dialects by overloading this function.
+        """
+        for dialect_name, dialect_factory in get_all_dialects().items():
+            self.register_dialect(dialect_name, dialect_factory)
```

Comment:
I am not entirely sure that anyone should inherit `MLContext`, so I would drop this comment.
Which is also why I'm not sure we should add this method here? Maybe we should just have `register_dialects` to make this simpler and just have `context.register_dialects(get_all_dialects())`? Or keep what we had before.

At least, I would prefer keep this a conscious decision for the user to add all dialects or not, and for that asking them to call `get_all_dialects` make sense to me if that makes sense? I don't know, what do you think?

Reply:
I think this gets especially weird for users of xDSL that want to add their own dialects, it seems a little cleaner to let them do their own combinations of different dialects rather than strongly encouraging inheritance in this way.